pkg/api/protobuf/go/minder/v1/api.go

```diff
@@ -0,0 +1,130 @@
+// SPDX-FileCopyrightText: Copyright 2024 The Minder Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package v1
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"github.com/mindersec/minder/internal/util/jsonyaml"
+)
+
+const (
+	// VersionV1 is the v1 version of most API resources
+	VersionV1 = "v1"
+)
+
+var (
+	// ErrInvalidResource is returned when an invalid resource is provided
+	ErrInvalidResource = fmt.Errorf("invalid resource")
+	// ErrInvalidResourceType is returned when an invalid resource type is provided
+	ErrInvalidResourceType = fmt.Errorf("invalid resource type")
+	// ErrNotAResource is returned when the provided resource is not a Minder resource
+	ErrNotAResource = fmt.Errorf("not a Minder resource")
+	// ErrResourceTypeMismatch is returned when the resource type does not match the provided resource
+	ErrResourceTypeMismatch = fmt.Errorf("resource type mismatch")
+)
+
+// ResourceType is the type of resource. Minder resources are all the objects that Minder manages.
+// They include policy resources (rule types, profiles, etc.), entity resources (repositories, artifacts, etc.),
+// and other resources like projects.
+type ResourceType string
+
+const (
+	// RuleTypeResource is a rule type resource
+	RuleTypeResource ResourceType = "rule-type"
+	// ProfileResource is a profile resource
+	ProfileResource ResourceType = "profile"
+	// EntityInstanceResource is an entity instance resource
+	EntityInstanceResource ResourceType = "entity-instance"
+	// RepositoryResource is a repository resource. Note that this
+	// will be deprecated in the future. In favor of EntityInstanceResource.
+	RepositoryResource ResourceType = "repository"
+	// ArtifactResource is an artifact resource. Note that this
+	// will be deprecated in the future. In favor of EntityInstanceResource.
+	ArtifactResource ResourceType = "artifact"
+	// ProjectResource is a project resource
+	ProjectResource ResourceType = "project"
+)
+
+// IsValid checks if the resource type is valid
+func (rt ResourceType) IsValid() bool {
+	switch rt {
+	case RuleTypeResource, ProfileResource, EntityInstanceResource, RepositoryResource, ArtifactResource,
+		ProjectResource:
+		return true
+	}
+	return false
+}
+
+var (
+	resourceMatchers = map[ResourceType]protoreflect.ProtoMessage{
+		RuleTypeResource:       &RuleType{},
+		ProfileResource:        &Profile{},
+		EntityInstanceResource: &EntityInstance{},
+		RepositoryResource:     &Repository{},
+		ArtifactResource:       &Artifact{},
+	}
+)
+
+// ResourceMeta defines the basic metadata for a resource within
+// Minder. This is used as a common interface and to determine
+// the type of resource.
+type ResourceMeta struct {
+	// Version is the version of the resource
+	Version string `json:"version" yaml:"version" validate:"required"`
```

Comment:
What kind of resources does this cover?  It looks like you want to cover both RuleTypes/Profiles (which would let us just `apply` a directory) _and_ entities / repositories / artifacts -- what's the use-case for those being on-disk, and do they apply in the same way?  Will we have up- and down-conversion on the server side, or is this simply detecting mismatches?

Given that proto can produce `Get` methods, what would you think of the following instead of `ParseResource`?

```golang

type ResourceMeta interface {
  protoreflect.ProtoMessage

  GetVersion() string
  GetType() string
  GetName() string
}

var _ ResourceMeta = (*RuleType)(nil)
var _ ResourceMeta = (*RuleType)(nil)
var _ ResourceMeta = (*EntityInstance)(nil)
var _ ResourceMeta = (*Repository)(nil)
var _ ResourceMeta = (*Artifact)(nil)

func ParseResource(r io.Reader, out ResourceMeta) error {
  // no need for io.TeeReader
  if err := json.NewDecoder(tee).Decode(out); err != nil {
    return fmt.Errorf("Error validating %T: %w", out, err)
  }
  // validate, including checking GetName() matches the type.
}
```

Reply:
Mostly this is useful for projects, rule types and profiles, but given thay we use the same format, I decided to apply it to entities as well. Trying to unify parsing and validation for ease of testing.

This suggestion is super handy, thanks! I'll try it tomorrow